### DISCUSSION TITLE
No crash in JIT? I was expecting one....

### DISCUSSION THREAD

**Discussion by jecampagne** [jax-ml/jax#15236] [Q&A]
Hi,
Although astonishing sometimes I am curious to know why a code is not crashing :) 

In the thread of **trace-time constant** and this [DOC](https://jax.readthedocs.io/en/latest/jep/4410-omnistaging.html), I was expecting a JIT crash for this example
```python
def exo2(x):
  n = x.shape[0]             # how n is know at compile time?
  print("x: ",x, type(x))
  jax.debug.print("x = {}",x)
  print("n: ",n, type(n))
  y = jnp.arange(0.,n)    # <-- crash expected here !
  return jnp.sum(x*y)
```
When no JIT is activated:
```python
x=jax.random.normal(jax.random.PRNGKey(10),shape=(10,))
exo2(x)
```
leads to 
```
x:  [ 0.4676388  -0.26684377  1.1685165  -0.48255268 -1.4658067   0.408608
 -0.12636082 -0.16064268  0.876103   -0.23290116] <class 'jaxlib.xla_extension.ArrayImpl'>
x = [ 0.4676388  -0.26684377  1.1685165  -0.48255268 -1.4658067   0.408608
 -0.12636082 -0.16064268  0.876103   -0.23290116]
n:  10 <class 'int'>
Array(-0.16760588, dtype=float32)
```

And I am curious to see that `jit(exo2)(x)`does not crash and leads to 
```
x:  Traced<ShapedArray(float32[10])>with<DynamicJaxprTrace(level=1/0)> <class 'jax.interpreters.partial_eval.DynamicJaxprTracer'>
n:  10 <class 'int'>
x = [ 0.4676388  -0.26684377  1.1685165  -0.48255268 -1.4658067   0.408608
 -0.12636082 -0.16064268  0.876103   -0.23290116]

Array(-0.16760588, dtype=float32)
```

I was (wrongly) expected that the `n` variable cannot be estimated at compile time ? I probably missed something.

**Comment by jakevdp** [comment]
Thanks for the question! Properties of arrays like shapes, sizes, and dtypes are always static within JAX, so they can be used to define other static quantities. Does that make sense?

**Reply by jecampagne** [reply]
Hum 
What is the difference then for this exemple
```python
def ex1(x):
  size = jnp.prod(jnp.array(x.shape))
  return jnp.sum(x.reshape((size,)))
```
to crash in the JIT version due to the reshape: => size computation starts with x.shape static but the jnp.prod not... Of course the solution is to use Pure Numpy operation here ie np.prod(np.array(x.shape)) , but hard to see the differnce with my exo2.

**Reply by jecampagne** [reply]
I am even more astonihed by the following example
```python
def ex0(x):
  size = np.prod(np.array(x.shape))
  return np.sum(x.reshape((size,)))
```
dealing with pure Numpy function
1) JIT is ok 
```python
 jit(ex0)(np.arange(0,12.,1).reshape((3, 4)))
```
gives 66. 
2) grad is ok too !!!
```python
grad(ex0)(np.arange(0,12.,1).reshape((3, 4)))
```
gives
```
Array([[1., 1., 1., 1.],
       [1., 1., 1., 1.],
       [1., 1., 1., 1.]], dtype=float32)
```
How is it possible that pure Numpy functions are ok here???? I'm completly lost.

**Reply by soraros** [reply]
Given what @jakevdp mentioned above
> Properties of arrays like shapes, sizes, and dtypes are always static within JAX

and and the fact that `jnp` functions (`jnp.reshape` in your case) which takes these properties as argument also require them to be static, we might annotate the your function like the follows:
```python
def ex1(x: Dynamic):
  shape: Static = x.shape
  size: Dynamic = jnp.prod(shape)
  new_shape = (size,): Dynamic

  x.reshape: Static -> Dynamic
  new_x = x.reshape(new_shape)  # mismatch here
  ...
```

**Reply by soraros** [reply]
@jakevdp This looks like a bug...
```python
>>> jax.make_jaxpr(np.sum)(np.arange(10))
{ lambda ; a:i64[10]. let b:i64[] = reduce_sum[axes=(0,)] a in (b,) }
```

Looks like NumPy being too polymorphic:
https://github.com/numpy/numpy/blob/e5d6c9be432363ddb150b03a336c97ea9d5b6920/numpy/core/fromnumeric.py#L71-L88
Is there anything we can do here?

**Reply by jecampagne** [reply]
> Looks like NumPy being too polymorphic here:
> https://github.com/numpy/numpy/blob/e5d6c9be432363ddb150b03a336c97ea9d5b6920/numpy/core/fromnumeric.py#L71-L88

Well if it is related to `ex0` what does this means?  It is a  bug? but how one can see that it is a bug? If I write a code  that accidentally uses this hidden failure how I can anticipate the failure? Of course I am happy to discover a bug that will be cured at next release but it is hard to follow and make some advise for the users. (ie I am writing some tutos for colleagues)

At least 
```python
def f(x):
  return np.exp(x)
``` 
failes to be jitted/grad ..... ouf :)

**Reply by jakevdp** [reply]
@jecampagne - regarding your question about the difference between `numpy` operaitons and `jax.numpy` operations, this might be useful reading: https://jax.readthedocs.io/en/latest/notebooks/thinking_in_jax.html#static-vs-traced-operations

@soraros - I don't think that is a bug. `np.sum` is a function that first looks for whether its input has a `sum` method, and if so, it calls it. When you call `make_jaxpr`, it will pass a JAX tracer to the function, and JAX tracers do have a `sum` method. So everything looks to be working as expected.

**Reply by jecampagne** [reply]
Thanks for the doc link. Does it comes from my questioning?

**Reply by soraros** [reply]
> ... So everything looks to be working as expected.

@jakevdp Yeah, I read the implementation. Still find the lost of "phase distinction" unsettling though.

This is from the doc:
```
A useful pattern is to use numpy for operations that should be static (i.e. done at compile-time),
and use jax.numpy for operations that should be traced (i.e. compiled and executed at run-time).
```

Should we mention using NumPy functions don't provide guarantee (say, raise when not possible at compile time) for staticness here?

**Reply by jakevdp** [reply]
One strength of that doc is its conciseness; I don't think adding a side-conversation about numpy implementation details would add much value.

**Reply by jecampagne** [reply]
Oh. yes for sure, I was asking if this doc has been triggered by our discussion? I'm glad if it is the case. Now I feel more comfortable to explain why there was NO crash... Thanks.

**Reply by jakevdp** [reply]
Hi @jecampagne – my answer was in response to @soraros. Regarding your question, no this doc is a few years old, but it's a frequent question so it's nice to have a good writeup to point to 😀
